File: rust/sort/counting_sort.rs
/// 计数排序 (Counting Sort)
///
/// # 算法原理
/// 计数排序是一种非比较排序算法：
/// 1. 找出待排序数组中的最大值和最小值
/// 2. 创建计数数组，统计每个值出现的次数
/// 3. 累加计数，得到每个值在输出数组中的位置
/// 4. 反向遍历原数组，将元素放到正确位置
///
/// # 时间复杂度: O(n + k)，其中k是数据范围
/// # 空间复杂度: O(n + k)
/// # 稳定性: 稳定（使用反向遍历时）
///
/// # 适用场景: 数据范围不大的整数排序

/// 计数排序 - 适用于非负整数
fn counting_sort(arr: &mut [usize]) {
    if arr.len() <= 1 {
        return;
    }

    // 找到最大值
    let max_val = *arr.iter().max().unwrap();

    // 创建计数数组
    let mut count = vec![0usize; max_val + 1];

    // 统计每个值的出现次数
    for &num in arr.iter() {
        count[num] += 1;
    }

    // 将排序后的值写回原数组
    let mut index = 0;
    for (val, &cnt) in count.iter().enumerate() {
        for _ in 0..cnt {
            arr[index] = val;
            index += 1;
        }
    }
}

/// 稳定版计数排序 - 适用于有符号整数
fn counting_sort_stable(arr: &mut [i32]) {
    if arr.len() <= 1 {
        return;
    }

    // 找到最大值和最小值
    let min_val = *arr.iter().min().unwrap();
    let max_val = *arr.iter().max().unwrap();

    // 计算偏移量和范围
    let range = (max_val - min_val + 1) as usize;

    // 创建计数数组
    let mut count = vec![0usize; range];

    // 统计每个值的出现次数
    for &num in arr.iter() {
        count[(num - min_val) as usize] += 1;
    }

    // 累加计数，得到每个值的结束位置
    for i in 1..range {
        count[i] += count[i - 1];
    }

    // 创建输出数组，反向遍历保证稳定性
    let mut output = vec![0i32; arr.len()];
    for &num in arr.iter().rev() {
        let idx = (num - min_val) as usize;
        count[idx] -= 1;
        output[count[idx]] = num;
    }

    // 复制回原数组
    arr.copy_from_slice(&output);
}

fn main() {
    println!("=== 计数排序 (Counting Sort) 演示 ===\n");

    // 非负整数排序
    let mut numbers: Vec<usize> = vec![4, 2, 2, 8, 3, 3, 1];
    println!("原始: {:?}", numbers);
    counting_sort(&mut numbers);
    println!("排序后: {:?}", numbers);

    println!();

    // 有符号整数排序
    let mut signed = vec![-5, 3, -1, 0, 3, -5, 2, 1];
    println!("有符号: {:?}", signed);
    counting_sort_stable(&mut signed);
    println!("排序后: {:?}", signed);

    println!();

    // 展示计数过程
    let mut demo: Vec<usize> = vec![1, 4, 1, 2, 7, 5, 2];
    println!("演示数组: {:?}", demo);

    let max_val = *demo.iter().max().unwrap();
    let mut count = vec![0usize; max_val + 1];
    for &num in demo.iter() {
        count[num] += 1;
    }
    println!("计数数组: {:?}", count);

    counting_sort(&mut demo);
    println!("排序结果: {:?}", demo);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_counting_sort() {
        let mut arr: Vec<usize> = vec![4, 2, 2, 8, 3, 3, 1];
        counting_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 2, 3, 3, 4, 8]);
    }

    #[test]
    fn test_counting_sort_stable() {
        let mut arr = vec![-5, 3, -1, 0, 3, -5, 2];
        counting_sort_stable(&mut arr);
        assert_eq!(arr, vec![-5, -5, -1, 0, 2, 3, 3]);
    }

    #[test]
    fn test_empty() {
        let mut arr: Vec<usize> = vec![];
        counting_sort(&mut arr);
        assert_eq!(arr, vec![]);
    }

    #[test]
    fn test_single() {
        let mut arr: Vec<usize> = vec![42];
        counting_sort(&mut arr);
        assert_eq!(arr, vec![42]);
    }

    #[test]
    fn test_all_same() {
        let mut arr: Vec<usize> = vec![5, 5, 5, 5];
        counting_sort(&mut arr);
        assert_eq!(arr, vec![5, 5, 5, 5]);
    }
}

File: cpp/counting_sort.cpp
/**
 * 计数排序 (Counting Sort)
 *
 * 算法原理：统计每个元素出现的次数，然后根据计数结果将元素放到正确位置。
 * 适用于整数排序，特别是范围较小的非负整数。
 *
 * 时间复杂度：O(n + k)，其中k是数据范围
 * 空间复杂度：O(n + k)
 * 稳定性：稳定
 */

#include <algorithm>
#include <iostream>
#include <vector>

/**
 * 计数排序（基本版本，只能处理非负整数）
 */
void countingSort(std::vector<int> &arr) {
  if (arr.empty())
    return;

  // 找到最大值
  int maxVal = *std::max_element(arr.begin(), arr.end());

  // 创建计数数组
  std::vector<int> count(maxVal + 1, 0);

  // 统计每个元素出现的次数
  for (int num : arr) {
    count[num]++;
  }

  // 根据计数重建数组
  int index = 0;
  for (int i = 0; i <= maxVal; i++) {
    while (count[i] > 0) {
      arr[index++] = i;
      count[i]--;
    }
  }
}

/**
 * 计数排序（稳定版本）
 */
void countingSortStable(std::vector<int> &arr) {
  if (arr.empty())
    return;

  int n = arr.size();
  int minVal = *std::min_element(arr.begin(), arr.end());
  int maxVal = *std::max_element(arr.begin(), arr.end());
  int range = maxVal - minVal + 1;

  std::vector<int> count(range, 0);
  std::vector<int> output(n);

  // 统计每个元素出现次数
  for (int num : arr) {
    count[num - minVal]++;
  }

  // 计算累积计数（确定每个元素的结束位置）
  for (int i = 1; i < range; i++) {
    count[i] += count[i - 1];
  }

  // 从后向前遍历（保证稳定性），将元素放入正确位置
  for (int i = n - 1; i >= 0; i--) {
    int pos = count[arr[i] - minVal] - 1;
    output[pos] = arr[i];
    count[arr[i] - minVal]--;
  }

  // 复制回原数组
  arr = output;
}

void printArray(const std::vector<int> &arr, const std::string &msg) {
  std::cout << msg;
  for (int num : arr)
    std::cout << num << " ";
  std::cout << std::endl;
}

int main() {
  std::cout << "========== 计数排序演示 ==========" << std::endl;
  std::vector<int> arr = {4, 2, 2, 8, 3, 3, 1, 7, 5, 6};
  printArray(arr, "排序前: ");
  countingSort(arr);
  printArray(arr, "排序后: ");

  std::cout << "\n========== 稳定计数排序 ==========" << std::endl;
  std::vector<int> arr2 = {4, 2, 2, 8, 3, 3, 1, 7, 5, 6};
  printArray(arr2, "排序前: ");
  countingSortStable(arr2);
  printArray(arr2, "排序后: ");

  std::cout << "\n测试有重复元素的数组：" << std::endl;
  std::vector<int> arr3 = {1, 4, 1, 2, 7, 5, 2};
  printArray(arr3, "排序前: ");
  countingSort(arr3);
  printArray(arr3, "排序后: ");

  return 0;
}
